### PR TITLE
Fixes OWASP/www-community#992

### PR DESCRIPTION
**Pull request by jakekarnes42**
This is a significant rewrite of the current [Insufficient Session-ID Length page](https://owasp.org/www-community/vulnerabilities/Insufficient_Session-ID_Length). 

The motivation for this rewrite is explained in #992. 

I've done my best to be careful with the formatting and linked images, but I didn't have a great way to validate this formatting is working correctly outside of the markdown preview. The images don't display in the preview, but I linked the new images in the same manner as the original page. 

Feedback is welcome!

Fixes OWASP/www-community#992

### CODE REVIEW

**Comment by kingthorin** [comment]
https://owasp.org/migration/ provides info on setting up a local testing environment.

**Comment by jakekarnes42** [comment]
Thank you for the quick reply. I've used the local environment as suggested to validate the formatting. I made a few improvements and I think the formatting is clear now. Here are two quick examples from some formatting-heavy sections:

### Example 1:
image

### Example 2:
image

File: pages/vulnerabilities/Insufficient_Session-ID_Length.md [hunk 1]
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
...
 
-Session identifiers should be at least 128 bits long to prevent brute-force session guessing attacks.
+Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. It's important to understand that session ID length alone is not a sufficient indicator of security. Instead, the critical factor is the entropy (or randomness) contained within the session identifier. While the length of the session ID can affect how much entropy is encoded, the actual entropy depends on how the session ID is generated. Encoding methods (e.g., hexadecimal, Base64, etc) influence how the session id is represented, but the security ultimately comes from the randomness of the identifier, not the encoding itself.
 
-The WebLogic deployment descriptor should specify a session identifier length of at least 128 bits. A shorter session identifier leaves the application open to brute-force session guessing attacks. If an attacker can guess an authenticated user's session identifier, they can take over the user's session. The remainder of this explanation will detail a back-of-the-envelope justification for a 128 bit session identifier.
+### Key Takeaways:
```
Comment (kingthorin) [inline pages/vulnerabilities/Insufficient_Session-ID_Length.md:18]: Drop the colon

File: pages/vulnerabilities/Insufficient_Session-ID_Length.md [hunk 2]
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
 
 ## Description
 
-Session identifiers should be at least 128 bits long to prevent brute-force session guessing attacks.
+Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. It's important to understand that session ID length alone is not a sufficient indicator of security. Instead, the critical factor is the entropy (or randomness) contained within the session identifier. While the length of the session ID can affect how much entropy is encoded, the actual entropy depends on how the session ID is generated. Encoding methods (e.g., hexadecimal, Base64, etc) influence how the session id is represented, but the security ultimately comes from the randomness of the identifier, not the encoding itself.
```
Comment (kingthorin) [inline pages/vulnerabilities/Insufficient_Session-ID_Length.md:16]: Could you ensure consistency in ID vs id vs Id?

This paragraph uses one and two. Chances are there are other places in the doc too.

IMHO full caps is the way to go, but consistency is king.

File: pages/vulnerabilities/Insufficient_Session-ID_Length.md [hunk 3]
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
...
+2. **Session ID using Base64 encoding**:
+    - Example Session ID #1: `abcdefghijkl`
+    - Example Session ID #2: `mnopqrstuvwx`
+    - This 12-character Base64 value can represent 72 bits of entropy. Base64 is more efficient than hexadecimal because each character represents more than 4 bits of entropy. The result is that even though this session ID is shorter than the previous hexadecimal example, it offers significantly more security.
+    - We can use the popular encoding/decoding tool CyberChef to confirm this value can represent a 72-bit number. We can see [example #1](https://gchq.github.io/CyberChef/#recipe=From_Base64('A-Za-z0-9%2B/%3D',true,false)To_Binary('None',8)&input=YWJjZGVmZ2hpamts) and [example #2](https://gchq.github.io/CyberChef/#recipe=From_Base64('A-Za-z0-9%2B/%3D',true,false)To_Binary('None',8)&input=bW5vcHFyc3R1dnd4) decode to a 72-bit binary value. 
+
+3. **16-byte hexadecimal session ID with fixed characters**:
+    - Example Session ID #1: `DECAFBAD12345678`
+    - Example Session ID #2: `DECAFBAD87654321`
+    - Example Session ID #3: `DECAFBAD12341234`
```
Comment (kingthorin) [inline pages/vulnerabilities/Insufficient_Session-ID_Length.md:63]: 🚫☕️👎 that's wonderful. I appreciate this bit of humour

File: pages/vulnerabilities/Insufficient_Session-ID_Length.md [hunk 4]
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
...
+In this scenario, we'll assume that the set of valid session identifiers is **not** changing throughout the attack. In the real world, this could occur when the attacker is able to exhaust all possibilities quickly (e.g., low entropy or many requests per second) or when targeting long-lived identifiers (e.g., API keys or tokens). When the set of valid session identifiers remains static, the attack becomes more efficient because the attacker can eliminate previously guessed invalid session IDs. This scenario follows the hypergeometric distribution, where the expected number of guesses is the average position of the first valid session ID among all possible values. In this case, the expected time for an attacker to guess a valid session ID is:
+
+![](../assets/images/Session_id_guessing_2.png)
+
+This formula accounts for the increasing probability of success with each guess as the pool of possible valid session IDs shrinks.
+
+### Example Calculations
+The following provides some example calcuations based on different attack scenarios. 
+
+#### Example 1:
```
Comment (kingthorin) [inline pages/vulnerabilities/Insufficient_Session-ID_Length.md:95]: I'd drop all the trails non colons from headings, they seem unnecessary

File: pages/vulnerabilities/Insufficient_Session-ID_Length.md [hunk 5]
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
...
+
+Using Formula 2 (static session IDs):
+
+![](../assets/images/Session_id_guessing_6.png)
+
+We can see that even though the attacker has more potential session ids to compromise, and is making many more requests per second, the 64 bits of entropy in the session id make it effectively unguessable.
 
 ## Risk Factors
 
 - Attackers that are try to obtain a valid session ID for [Session hijacking](../attacks/Session_hijacking_attack).
```
Comment (kingthorin) [inline pages/vulnerabilities/Insufficient_Session-ID_Length.md:132]: trying

I know that's an existing typo but you may as well fix it

**Comment by jakekarnes42** [comment]
Thank you for the thoughtful feedback! My latest commit addresses those issues along with 2 minor content changes:

1. I added a brief explanation of what is meant by "bits of entropy" at the beginning. The first paragraph now reads:
>Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. Entropy refers to the amount of randomness or unpredictability in a value. Each "bit" of entropy doubles the number of possible outcomes, meaning a session ID with 64 bits of entropy can have 2<sup>64</sup> possible values. This randomness provides security, as it makes it harder for attackers to guess valid session IDs

2. I updated the content under the "Focus on Hexadecimal Encoding" heading. This paragraph is now:
> Hexadecimal encoding is commonly used in session identifiers. Each hexadecimal character (0-9, A-F) represents 4 bits of entropy. This means that to achieve 64 bits of entropy, a session ID must contain at least 16 random hexadecimal characters, which provides 2<sup>64</sup> possible combinations. When each character is chosen randomly, the session ID offers the desired 64 bits of entropy, making it secure against brute-force guessing attacks. Let’s explore good and bad examples of hexadecimal session IDs to better understand this:

The second update makes the examples slightly redundant, but seeing as this is a source of confusion for some people, I think it's worth being extra clear.

**Review by kingthorin — APPROVED** [review]
Thanks!

**Comment by kingthorin** [comment]
Now live: https://owasp.org/www-community/vulnerabilities/Insufficient_Session-ID_Length

**Pull request merged**